File: JsWs/apps/lan-streamer/api/src/audio/ws-handlers.ts
import { randomUUID } from "node:crypto";
import type { IncomingMessage } from "node:http";
import type { Duplex } from "node:stream";
import { WebSocketServer, type WebSocket } from "ws";
import { accountForToken } from "../auth.js";
import { addActiveHost, getStream, removeActiveHost } from "../store.js";
import type { HostAudioSource } from "../types.js";
import {
  BYTES_PER_FRAME,
  forwardRawFrame,
  forwardSimpleFrame,
  pushHostFrame,
  registerHost,
  registerListener,
  registerRawListener,
  registerSimpleListener,
  unregisterHost,
  unregisterListener,
  unregisterRawListener,
  unregisterSimpleListener,
} from "./mixer.js";

const wss = new WebSocketServer({ noServer: true });

// Tracks which host connection is currently "live" on each simple/raw stream
// - a second host starting up supersedes whoever's currently in the map for
// that streamId, rather than mixing with them. Separate maps per mode since
// "simple" and "raw" are otherwise-independent stream namespaces.
const activeSimpleHosts = new Map<string, { connectionId: string; ws: WebSocket }>();
const activeRawHosts = new Map<string, { connectionId: string; ws: WebSocket }>();

function closeWith(socket: Duplex, code: number, reason: string): void {
  socket.end(`HTTP/1.1 ${code} ${reason}\r\n\r\n`);
}

export function handleUpgrade(req: IncomingMessage, socket: Duplex, head: Buffer): void {
  const url = new URL(req.url ?? "", "http://localhost");
  const [, , kind, streamId] = url.pathname.split("/"); // /ws/host/:id or /ws/listen/:id
  const token = url.searchParams.get("token") ?? undefined;

  const account = accountForToken(token);
  if (!account) {
    closeWith(socket, 401, "Unauthorized");
    return;
  }

  const stream = streamId ? getStream(streamId) : undefined;
  if (!stream) {
    closeWith(socket, 404, "Not Found");
    return;
  }

  if (kind === "host") {
    if (stream.ownerAccountId !== account.userId) {
      closeWith(socket, 403, "Forbidden");
      return;
    }
    // An account can host any number of distinct streams at once (each from a
    // different device/window "device stream"); only a single connection is
    // ever tied to one streamId, since that's inherent to one WebSocket.
    const sourceParam = url.searchParams.get("source");
    const audioSource: HostAudioSource =
      sourceParam === "system" || sourceParam === "test-tone" ? sourceParam : "microphone";
    wss.handleUpgrade(req, socket, head, (ws) => {
      if (stream.mode === "simple") attachExclusiveHost(ws, streamId, account.userId, audioSource, activeSimpleHosts, forwardSimpleFrame);
      else if (stream.mode === "raw") attachExclusiveHost(ws, streamId, account.userId, audioSource, activeRawHosts, (id, _accountId, frame) => forwardRawFrame(id, frame));
      else attachHost(ws, streamId, account.userId, audioSource);
    });
    return;
  }

  if (kind === "listen") {
    wss.handleUpgrade(req, socket, head, (ws) => {
      if (stream.mode === "simple") attachSimpleListener(ws, streamId);
      else if (stream.mode === "raw") attachRawListener(ws, streamId);
      else attachListener(ws, streamId);
    });
    return;
  }

  closeWith(socket, 404, "Not Found");
}

function attachHost(ws: WebSocket, streamId: string, accountId: number, audioSource: HostAudioSource): void {
  const connectionId = randomUUID();

  registerHost(streamId, connectionId, accountId);
  addActiveHost(streamId, connectionId, accountId, audioSource);

  ws.on("message", (data, isBinary) => {
    if (!isBinary) return;
    const frame = Buffer.isBuffer(data) ? data : Buffer.from(data as ArrayBuffer);
    if (frame.length !== BYTES_PER_FRAME) return;
    pushHostFrame(streamId, connectionId, frame);
  });

  ws.on("close", () => {
    unregisterHost(streamId, connectionId);
    removeActiveHost(streamId, connectionId, accountId);
  });
}

function attachListener(ws: WebSocket, streamId: string): void {
  registerListener(streamId, ws);
  ws.on("close", () => unregisterListener(streamId, ws));
}

// Shared by "simple" and "raw" modes: both allow exactly one active host at
// a time, with a second host superseding whoever was streaming before it -
// they only differ in what happens to each frame once it arrives (the
// `forward` callback), passed in by the caller along with which mode's
// active-host map to track this connection in.
function attachExclusiveHost(
  ws: WebSocket,
  streamId: string,
  accountId: number,
  audioSource: HostAudioSource,
  activeHosts: Map<string, { connectionId: string; ws: WebSocket }>,
  forward: (streamId: string, accountId: number, frame: Buffer) => void,
): void {
  const connectionId = randomUUID();

  // A second host starting up on the same stream supersedes whoever was
  // streaming on it - tell that previous connection why it's being cut off
  // (rather than leaving it to guess from a plain close) and disconnect it,
  // instead of letting two sources collide.
  const previous = activeHosts.get(streamId);
  if (previous && previous.connectionId !== connectionId) {
    if (previous.ws.readyState === previous.ws.OPEN) {
      previous.ws.send(JSON.stringify({ type: "superseded" }));
    }
    previous.ws.close();
  }
  activeHosts.set(streamId, { connectionId, ws });

  addActiveHost(streamId, connectionId, accountId, audioSource);

  ws.on("message", (data, isBinary) => {
    if (!isBinary) return;
    // Ignore frames from a connection that's since been superseded but
    // hasn't finished closing yet - only the current active host forwards.
    if (activeHosts.get(streamId)?.connectionId !== connectionId) return;
    const frame = Buffer.isBuffer(data) ? data : Buffer.from(data as ArrayBuffer);
    if (frame.length !== BYTES_PER_FRAME) return;
    forward(streamId, accountId, frame);
  });

  ws.on("close", () => {
    if (activeHosts.get(streamId)?.connectionId === connectionId) {
      activeHosts.delete(streamId);
    }
    removeActiveHost(streamId, connectionId, accountId);
  });
}

function attachSimpleListener(ws: WebSocket, streamId: string): void {
  registerSimpleListener(streamId, ws);
  ws.on("close", () => unregisterSimpleListener(streamId, ws));
}

function attachRawListener(ws: WebSocket, streamId: string): void {
  registerRawListener(streamId, ws);
  ws.on("close", () => unregisterRawListener(streamId, ws));
}

File: JsWs/apps/lan-streamer/api-rs/src/ws_handlers.rs
use axum::extract::ws::{Message, WebSocket, WebSocketUpgrade};
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use serde::Deserialize;
use uuid::Uuid;

use crate::auth::account_for_token;
use crate::mixer::{ExclusiveHost, BYTES_PER_FRAME};
use crate::types::{HostAudioSource, StreamMode};
use crate::AppState;

#[derive(Deserialize)]
pub struct HostQuery {
    token: Option<String>,
    source: Option<String>,
}

#[derive(Deserialize)]
pub struct ListenQuery {
    token: Option<String>,
}

pub async fn ws_host(
    Path(stream_id): Path<String>,
    Query(query): Query<HostQuery>,
    State(state): State<AppState>,
    ws: WebSocketUpgrade,
) -> Response {
    let Some(account) = query.token.as_deref().and_then(|t| account_for_token(&state.store, t)) else {
        return (StatusCode::UNAUTHORIZED, "Unauthorized").into_response();
    };
    let Some(stream) = state.store.get_stream(&stream_id) else {
        return (StatusCode::NOT_FOUND, "Not Found").into_response();
    };
    if stream.owner_account_id != account.user_id {
        return (StatusCode::FORBIDDEN, "Forbidden").into_response();
    }

    // An account can host any number of distinct streams at once (each from
    // a different device/window "device stream"); only a single connection
    // is ever tied to one streamId, since that's inherent to one WebSocket.
    let audio_source = match query.source.as_deref() {
        Some("system") => HostAudioSource::System,
        Some("test-tone") => HostAudioSource::TestTone,
        _ => HostAudioSource::Microphone,
    };

    ws.on_upgrade(move |socket| async move {
        match stream.mode {
            StreamMode::Simple => attach_simple_host(socket, stream_id, account.user_id, audio_source, state).await,
            StreamMode::Raw => attach_raw_host(socket, stream_id, account.user_id, audio_source, state).await,
            StreamMode::Merged => attach_host(socket, stream_id, account.user_id, audio_source, state).await,
        }
    })
}

pub async fn ws_listen(Path(stream_id): Path<String>, Query(query): Query<ListenQuery>, State(state): State<AppState>, ws: WebSocketUpgrade) -> Response {
    let Some(_account) = query.token.as_deref().and_then(|t| account_for_token(&state.store, t)) else {
        return (StatusCode::UNAUTHORIZED, "Unauthorized").into_response();
    };
    let Some(stream) = state.store.get_stream(&stream_id) else {
        return (StatusCode::NOT_FOUND, "Not Found").into_response();
    };

    ws.on_upgrade(move |socket| async move {
        match stream.mode {
            StreamMode::Simple => attach_simple_listener(socket, stream_id, state).await,
            StreamMode::Raw => attach_raw_listener(socket, stream_id, state).await,
            StreamMode::Merged => attach_listener(socket, stream_id, state).await,
        }
    })
}

async fn attach_host(mut socket: WebSocket, stream_id: String, account_id: i64, audio_source: HostAudioSource, state: AppState) {
    let connection_id = Uuid::new_v4().to_string();
    state.mixer.register_host(&stream_id, &connection_id, account_id);
    state.store.add_active_host(&stream_id, connection_id.clone(), account_id, audio_source).await;

    while let Some(Ok(msg)) = socket.recv().await {
        if let Message::Binary(data) = msg {
            if data.len() == BYTES_PER_FRAME {
                state.mixer.push_host_frame(&stream_id, &connection_id, data);
            }
        }
    }

    state.mixer.unregister_host(&stream_id, &connection_id);
    state.store.remove_active_host(&stream_id, &connection_id, account_id).await;
}

async fn attach_listener(mut socket: WebSocket, stream_id: String, state: AppState) {
    let connection_id = Uuid::new_v4().to_string();
    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
    state.mixer.register_listener(&stream_id, connection_id.clone(), tx);

    loop {
        tokio::select! {
            frame = rx.recv() => {
                match frame {
                    Some(bytes) => { if socket.send(Message::Binary(bytes.to_vec())).await.is_err() { break; } }
                    None => break,
                }
            }
            msg = socket.recv() => {
                if !matches!(msg, Some(Ok(_))) { break; }
            }
        }
    }

    state.mixer.unregister_listener(&stream_id, &connection_id);
}

async fn attach_simple_listener(mut socket: WebSocket, stream_id: String, state: AppState) {
    let connection_id = Uuid::new_v4().to_string();
    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
    state.mixer.register_simple_listener(&stream_id, connection_id.clone(), tx);

    loop {
        tokio::select! {
            frame = rx.recv() => {
                match frame {
                    Some(bytes) => { if socket.send(Message::Binary(bytes.to_vec())).await.is_err() { break; } }
                    None => break,
                }
            }
            msg = socket.recv() => {
                if !matches!(msg, Some(Ok(_))) { break; }
            }
        }
    }

    state.mixer.unregister_simple_listener(&stream_id, &connection_id);
}

async fn attach_raw_listener(mut socket: WebSocket, stream_id: String, state: AppState) {
    let connection_id = Uuid::new_v4().to_string();
    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
    state.mixer.register_raw_listener(&stream_id, connection_id.clone(), tx);

    loop {
        tokio::select! {
            frame = rx.recv() => {
                match frame {
                    Some(bytes) => { if socket.send(Message::Binary(bytes.to_vec())).await.is_err() { break; } }
                    None => break,
                }
            }
            msg = socket.recv() => {
                if !matches!(msg, Some(Ok(_))) { break; }
            }
        }
    }

    state.mixer.unregister_raw_listener(&stream_id, &connection_id);
}

// "simple" and "raw" hosting share the same single-active-host/superseding
// behavior - a second host starting up on the same stream supersedes
// whoever was streaming on it. They're written as two near-identical
// functions (rather than one generic one) since the two modes use separate
// active-host maps and forward functions, and the coordination logic here
// is already fiddly enough without adding generics on top.

async fn attach_simple_host(mut socket: WebSocket, stream_id: String, account_id: i64, audio_source: HostAudioSource, state: AppState) {
    let connection_id = Uuid::new_v4().to_string();
    let (close_tx, mut close_rx) = tokio::sync::mpsc::unbounded_channel::<()>();

    {
        let mut hosts = state.mixer.active_simple_hosts.lock().unwrap();
        if let Some(previous) = hosts.get(&stream_id) {
            let _ = previous.close.send(());
        }
        hosts.insert(stream_id.clone(), ExclusiveHost { connection_id: connection_id.clone(), close: close_tx });
    }
    state.store.add_active_host(&stream_id, connection_id.clone(), account_id, audio_source).await;

    loop {
        tokio::select! {
            msg = socket.recv() => {
                match msg {
                    Some(Ok(Message::Binary(data))) => {
                        let still_active = state.mixer.active_simple_hosts.lock().unwrap().get(&stream_id).map(|h| h.connection_id == connection_id).unwrap_or(false);
                        if still_active && data.len() == BYTES_PER_FRAME {
                            state.mixer.forward_simple_frame(&stream_id, account_id, &data);
                        }
                    }
                    Some(Ok(_)) => {}
                    Some(Err(_)) | None => break,
                }
            }
            _ = close_rx.recv() => {
                let _ = socket.send(Message::Text(r#"{"type":"superseded"}"#.into())).await;
                break;
            }
        }
    }

    {
        let mut hosts = state.mixer.active_simple_hosts.lock().unwrap();
        if hosts.get(&stream_id).map(|h| h.connection_id == connection_id).unwrap_or(false) {
            hosts.remove(&stream_id);
        }
    }
    state.store.remove_active_host(&stream_id, &connection_id, account_id).await;
}

async fn attach_raw_host(mut socket: WebSocket, stream_id: String, account_id: i64, audio_source: HostAudioSource, state: AppState) {
    let connection_id = Uuid::new_v4().to_string();
    let (close_tx, mut close_rx) = tokio::sync::mpsc::unbounded_channel::<()>();

    {
        let mut hosts = state.mixer.active_raw_hosts.lock().unwrap();
        if let Some(previous) = hosts.get(&stream_id) {
            let _ = previous.close.send(());
        }
        hosts.insert(stream_id.clone(), ExclusiveHost { connection_id: connection_id.clone(), close: close_tx });
    }
    state.store.add_active_host(&stream_id, connection_id.clone(), account_id, audio_source).await;

    loop {
        tokio::select! {
            msg = socket.recv() => {
                match msg {
                    Some(Ok(Message::Binary(data))) => {
                        let still_active = state.mixer.active_raw_hosts.lock().unwrap().get(&stream_id).map(|h| h.connection_id == connection_id).unwrap_or(false);
                        if still_active && data.len() == BYTES_PER_FRAME {
                            state.mixer.forward_raw_frame(&stream_id, &data);
                        }
                    }
                    Some(Ok(_)) => {}
                    Some(Err(_)) | None => break,
                }
            }
            _ = close_rx.recv() => {
                let _ = socket.send(Message::Text(r#"{"type":"superseded"}"#.into())).await;
                break;
            }
        }
    }

    {
        let mut hosts = state.mixer.active_raw_hosts.lock().unwrap();
        if hosts.get(&stream_id).map(|h| h.connection_id == connection_id).unwrap_or(false) {
            hosts.remove(&stream_id);
        }
    }
    state.store.remove_active_host(&stream_id, &connection_id, account_id).await;
}
